src/Xamarin.CodeAnalysis/RequiresSuperAttributeAnalyzer.cs

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+//
+// RequiresSuperAttributeAnalyzer.cs:
+//
+// Authors:
+//   Madeline Zhang  <v-madzh@microsoft.com>
+//
+// Copyright 2020 Xamarin Inc.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Text.RegularExpressions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace RequiresSuperAttribute
+{
+	/// <summary>
+	/// Analyzer that reports diagnostics for all methods with the [RequiresSuper] attribute which are overridden without a base.[methodname]() call in the method body
+	/// </summary>
+	[DiagnosticAnalyzer (LanguageNames.CSharp)]
+	public class RequiresSuperAttributeAnalyzer : DiagnosticAnalyzer
+	{
+		public static readonly string DiagnosticId = "XI0004";
+
+		static readonly DiagnosticDescriptor Rule =
+		new DiagnosticDescriptor (
+			DiagnosticId,
+			"Add base method call to methods with the [RequiresSuper] attribute",
+			"Overriding the {0} method requires a call to the overridden method",
+			"Notifications",
+			DiagnosticSeverity.Info,
+			isEnabledByDefault: true
+			);
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create (Rule);
+
+		const string RequiresSuperAttributeName = "RequiresSuperAttribute";
+		const string RequiresSuperAttributeNamespace = "ObjCRuntime";
+
+		public override void Initialize (AnalysisContext context)
+		{
+			context.ConfigureGeneratedCodeAnalysis (GeneratedCodeAnalysisFlags.None);
+			context.EnableConcurrentExecution ();
+			context.RegisterOperationBlockAction ((codeBlockContext) => {
+				// returns if the code block isn't an overridden method
+				var method = (IMethodSymbol)codeBlockContext.OwningSymbol;
+				if (method == null || !method.IsOverride) { return; }
+
+				// seeing if the [RequiresSuper] attribute is there
+				var baseType = method.ContainingType.BaseType;
+				ISymbol baseMethod = baseType.GetMembers (method.Name).FirstOrDefault ();
+				// returns if the [RequiresSuper] attribute isn't present
+				if (!baseMethod.GetAttributes ().Any (attr => (attr.AttributeClass.Name == RequiresSuperAttributeName && attr.AttributeClass.ContainingNamespace.Name == RequiresSuperAttributeNamespace))) {
+					return;
+				}
+
+				foreach (var block in codeBlockContext.OperationBlocks.Where ((b) => b.Kind == OperationKind.Block)) // loops once per method
```

Comment:
Comments on loops are 💯

Reply:
Filtering the array with `Where` is really neat as well!